File: 2473-max-sum-of-a-pair-with-equal-sum-of-digits/2473-max-sum-of-a-pair-with-equal-sum-of-digits.cpp
class Solution {
public:
    int digitSum(int n) {
        int sum = 0;
        while (n != 0) {
            sum += n % 10;
            n /= 10;
        }
        return sum;
    }

    int maximumSum(vector<int>& nums) {
        unordered_map<int, int> mp;
        int maxSum = -1;

        for (int i = 0; i < nums.size(); i++) {
            int digiSum = digitSum(nums[i]);
            if (mp.count(digiSum)) {
                maxSum = max(maxSum, mp[digiSum] + nums[i]);
                if (mp[digiSum] < nums[i])
                    mp[digiSum] = nums[i];
            } else mp[digiSum] = nums[i];
        }

        return maxSum;
    }
};

File: 2473-max-sum-of-a-pair-with-equal-sum-of-digits/2473-max-sum-of-a-pair-with-equal-sum-of-digits.go
func digitSum(num int) int {
    sum := 0
    for num != 0 {
        sum += num % 10
        num /= 10
    }
    return sum
}

func maximumSum(nums []int) int {
    mp := make(map[int]int)
    maxSum := -1

    for _, num := range nums {
        digiSum := digitSum(num)
        if oldVal, exists := mp[digiSum]; exists {
            maxSum = max(maxSum, num + oldVal);
            if oldVal < num {
                mp[digiSum] = num
            }
        } else {
            mp[digiSum] = num
        }
    }

    return maxSum
}